paddle/fluid/framework/parallel_executor.cc

```diff
@@ -0,0 +1,415 @@
+struct ComputationOpHandle : public OpHandle {
+  std::unique_ptr<OperatorBase> op_;
+
+  explicit ComputationOpHandle(const OpDesc &op_desc)
+      : op_(framework::OpRegistry::CreateOp(op_desc)) {}
+};
+
+struct ScaleLossGradOpHandle : public OpHandle {};
+
+struct NCCLAllReduceOpHandle : public OpHandle {};
+
+class ParallelExecutorPrivate {
+ public:
+  explicit ParallelExecutorPrivate(size_t num_threads = 12)
+      : pool_(num_threads) {}
+
+  std::unordered_map<platform::Place, Scope *, platform::PlaceHash>
+      local_scopes_;
+
+#ifdef PADDLE_WITH_CUDA
+  struct NCCLContext {
+    std::unique_ptr<platform::CUDADeviceContext> ctx_;
+    ncclComm_t comm;
+
+    explicit NCCLContext(int dev_id) {
+      ctx_.reset(new platform::CUDADeviceContext(platform::CUDAPlace(dev_id)));
+    }
+
+    cudaStream_t stream() const { return ctx_->stream(); }
+
+    int device_id() const {
+      return boost::get<platform::CUDAPlace>(ctx_->GetPlace()).device;
+    }
+
+    static void InitNCCLContext(std::map<int, NCCLContext> &contexts) {
+      std::vector<ncclComm_t> comms;
+      std::vector<int> devs;
+      comms.resize(contexts.size());
+      devs.reserve(contexts.size());
+
+      for (auto &ctx : contexts) {
+        devs.push_back(ctx.first);
+      }
+
+      NCCL_INVOKE(platform::dynload::ncclCommInitAll(
+          &comms[0], static_cast<int>(contexts.size()), &devs[0]));
+
+      int i = 0;
+      for (auto &ctx : contexts) {
+        ctx.second.comm = comms[i++];
+      }
+    }
+  };
+
+  std::map<int, NCCLContext> communication_streams_;
+
+  NCCLContext &GetNCCLCtx(platform::Place p) {
+    int dev_id = boost::get<platform::CUDAPlace>(p).device;
+    return communication_streams_.at(dev_id);
+  }
+
+#endif
+
+  platform::Place main_place_;
+
+  std::unordered_map<platform::Place,
+                     std::unordered_map<std::string, std::map<int, VarHandle>>,
+                     platform::PlaceHash>
+      vars_;
+  std::vector<std::unique_ptr<OpHandle>> ops_;
+
+  ThreadPool pool_;
+};
+
+// TODO(yy): Move this function somewhere
+ncclDataType_t ToNCCLDataType(std::type_index type) {
+  // FIXME!!
+  return ncclFloat;
+}
+
+ParallelExecutor::ParallelExecutor(
+    const std::vector<platform::Place> &places,
+    const std::unordered_set<std::string> &params,
+    const ProgramDesc &startup_program, const ProgramDesc &main_program,
+    const std::string &loss_var_name, Scope *scope)
+    : member_(new ParallelExecutorPrivate()) {
+  // Step 1. RunStartupProgram and Bcast the params to devs.
+  Executor exe(places[0]);
+  exe.Run(startup_program, scope, 0);
+  // Create local scopes
+  for (auto &place : places) {
+    member_->local_scopes_[place] = &scope->NewScope();
+  }
+  member_->main_place_ = places[0];
+
+  // Bcast Parameters to all GPUs
+  if (platform::is_gpu_place(member_->main_place_) &&
+      member_->local_scopes_.size() != 1) {  // Is CUDA
+    BuildNCCLCommunicator();
+    BCastParamsToGPUs(startup_program);
+  }
```

Comment:
Why do not initialize parameters on the respective devices？

Reply:
Since randomize result might not same when seed = 0